[Jay, 조이Zoey]  빨간 우체통 마을 찾기

altmit opened the pull request:
## 빨간 우체통 마을 찾기
안녕하세요 Alan! 이번에 처음으로 코드 리뷰 요청 드리게된 Jay, 조이입니다! 😊 이번 미션은 처음으로 페어 프로그래밍을 해보고, 알고리즘적인 부분이 많아서 구현 과정에 시행착오가 있었지만, 미션 요구사항을 최대한 충족시키기 위해서 class static 변수, DFS, 비동기, 재귀 등 학습하면서 코드에 녹여보려고 노력했습니다! 첫 리뷰 요청이라 설레네요 잘부탁드립니다 :)

## 주요  작업

### 💡마을 지도

- [x] 화면이 열리면, 기획서와 같은 모습의 마을들이 화면에 자동으로 만들어진다.
- [x] 마을의 개수, 위치, 크기는 랜덤하게 정해진다.
- [x] 마을안에 마을은 중첩되어 N개가 위치할 수 있다.
  - [x] 마을 안에 마을은 무한대로 존재할 수 있다. 이를 해결할 수 있는 해법을 사용한다.
- [x] 마을은 서로 위치가 겹칠 수가 없다.
- [x] 각 마을 안에는 우체통이 있거나 없다.
- [x] 우체통의 개수와 크기도 랜덤하게 생성된다.
  - [x] 우체통의 크기는 각각 다르다.

### 💡빨간 우체통 확인 기능

- [x] 만들어진 마을의 우체통을 탐색한 결과를 보여준다.
- [x] 빨간 우체통 확인 버튼은 클릭 후 빨간 우체통을 확인하는 기능이 2초 후에 동작한다.
- [x] [빨간 우체통 확인] 버튼을 클릭하면 우체통을 가진 마을을 탐색하는 작업을 시작한다.
  - [x] Node를 탐색하기 위한 API를 직접 개발한다. querySelector, querySelectorAll, getElementById
- [x] 빨간 우체통을 가진 마을 정보가 표시되고, 우체통 크기 순으로 정렬해서 정보를 표시한다. ex) D, I, J, K, L 총 5개의 마을에 빨간 우체통이 있습니다. 우체통의 크기는 L, K, J, I, D 순입니다.
  - [x] 우체통 크기 정렬을 위해 정렬 알고리즘을 직접 구현하여 이를 활용하여 정렬한다.
- [x] 빨간 우체통이 있는 마을의 테두리는 2초 뒤에 빨간색으로 변경된다.
- [x] setTimeout을 활용해서 지연처리를 하고, Promise 패턴을 적용한다.


<details>
<summary>동영상</summary>

https://user-images.githubusercontent.com/41321198/229013981-6b72f5c0-db99-4054-b696-60634cc5fcbd.mov

</details>

## 기술 키워드
- 병합 정렬
- class static 변수
- DFS
- 비동기
- 재귀 함수
- HTMLCollection

## 고민과 해결
- **[고민] 마을끼리 어떻게 겹치지 않게 만들 수 있을까?**
    - 이전에 만들어진 마을의 위치와 비교하여 상하좌우에 위치하는지 판단
    
- **[고민] 마을 안에 마을을 무한대로 생성할 수 있게 만들되, 이를 어떻게 해결할 수 있을까?**
    - Town 클래스의 children 프로퍼티에 마을을 만드는 함수를 실행한 결과를 담는다.
    - Town 너비와 높이 최소 길이를 정하여 부모 마을 대비 자식 마을의 변의 길이가 최소 길이보다 작아지면 함수를 실행하지 않고 리턴한다.
    
- **[고민] 부모 마을 내에 자식 마을을 몇개까지 만들어주어야 할까?**
    - 시도: 자식 마을 개수를 먼저 정하고, 부모 마을 영역에 해당 개수의 마을을 넣어 준다.
        - 로직에서 랜덤으로 정하는 가로, 세로, 위치 범위를 조절하지 않아 해당 개수가 될 때까지 너무 많은 반복이 필요했다.
    - 3회라는 기준을 만들고, 3회의 중첩이 일어나면 더 이상 마을을  만들지 않고 지금까지 만들어진 마을까지만 반환한다.
    - 랜덤 범위가 너무 커서 마을의 개수가 너무 적었다.
        - 랜덤의 범위를 수정하며 이 문제를 해결했다.
    
- **[고민] Node를 탐색하기 위한 API를 어떻게 구현할까?**
    - DFS pre-order
   
- **[고민] HTMLCollection 왜 forEach가 안될까?**
  - > GPT🤖 : Because HTMLCollection is live, it's not as simple to implement the forEach method as it is for a regular array. The forEach method requires that the entire collection be iterated over at the time the method is called, but this could cause problems if the collection is being updated in real-time.
That's why HTMLCollection doesn't support the forEach method, but you can still loop through its elements using a regular for loop or the for...of loop, as I demonstrated in my previous answer.

## 부가 설명
1. `rootInit()` : 초기 화면의 사이즈를 정해줍니다.
2. `makeVillage()` :  마을들의 구조를 생성합니다.
    1. 100 ~ (부모의 사이즈/2) -10 까지의 가로,세로 길이를 랜덤으로 뽑습니다.
    2. 10 ~ (부모의 사이즈 - (가로 or 세로) 사이즈 -20) 까지의 랜덤 top, left를 뽑습니다.
        1. 10, 20의 를 빼준 이유는 부모와 자식의  간격을 최소 10px을 주기 위해서 입니다.
    3. 최대 10개 까지 만들며 형제 노드와 비교해 겹치는 경우가 3번이 생기면  더 이상 마을을 만드는 것을 중단 합니다.
    4. 재귀로 위  과정을 반복하여 마을에 child 마을을 만듭니다.
3. `makeVillageView()` : 마을 데이터를 시각화 합니다.
    1. `makeVillage()`에서 만든 데이터를 기준으로  모두  화면에 렌더링 합니다.
4. `buttonClickEvent()` : 버튼 클릭 event를 추가합니다.

youzysu committed:
chore: 프리티어 설정 파일 생성

altmit committed:
빨간 우체통 마을 찾기 계획 작성

youzysu committed:
style: html 마크업 & css 스타일링

youzysu committed:
feat: 마을의 개수, 위치, 크기 정보를 정하는 함수 구현

- 부모의 너비와 높이를 인자로 받아 너비, 높이를 랜덤으로 정하고 우체통 여부를 랜덤으로 결정한다.
- 부모의 크기 내에서 만들 마을의 개수를 랜덤으로 정하고, 생성된 마을을 반환한다.
- 형제 마을과 서로 겹치지 않도록 확인한다.

youzysu committed:
feat: 마을 정보를 받아 화면에 렌더링한다.

youzysu committed:
feat: 자식 마을 생성 및 렌더링

altmit committed:
feat: 마을에 우체통 랜더링

youzysu committed:
refactor: 마을 만드는 로직 전반 리팩터링

- 함수 분리
- 모듈 분리

altmit committed:
feat: 우체통 크기를 랜덤하게 생성하여 dataset에 부여한다

altmit committed:
docs: readme 업데이트

altmit committed:
feat: 생성된 마을에 알파벳 순으로 이름을 부여한다

youzysu committed:
refactor: 응집도가 낮은 함수 개선

youzysu committed:
feat: DOM API 객체 생성

altmit committed:
refactor: domAPI  불필요한 반복 제거

altmit committed:
refactor: 로직 전반적인 부분 개선

- 부모와 겹치지 않는 경우에만 마을 프로퍼티 설정
- static 변수를 정의하여 마을 인스턴스의 데이터 관리
- 가독성을 위한 전반 수정
- Town class, utils 모듈화

gunoc commented on Town.js:
```diff
@@ -0,0 +1,61 @@
+    this.pointX = this.makeRandomPoint(parentWidth, this.width);
+    this.pointY = this.makeRandomPoint(parentHeight, this.height);
+    this.children = [];
+    this.hasMailbox = Math.random() < 0.4;
```
0.4인 이유가 있다면 무엇인지 궁금합니다..! 제한해두신 확률인 걸까요?

youzysu replied:
확률적으로 10개 중에 4개 정도의 우체통이 만들어졌을 때 DOM 탐색 시에 적당한 수가 만들어진다고 판단했어요. 처음에 50% 비율로 생성했을 때 생각보다 많이 생성이 돼서 가장 적당한 확률이었습니다!

gunoc commented on domAPI.js:
```diff
@@ -0,0 +1,74 @@
+      return nodeList;
+    }
+
+    return nodeList.filter(filter);
```
.filter메서드와 filter가 겹쳐 어떤 필터링을 하는지 모호하게 느껴지는 것 같습니다 이름을 체크하는 filter라면 역할을 고려하여 filter의 네이밍을 조금 더 구체화하면 어떨까요?

youzysu replied:
DOM 탐색을 여러번 다른 조건에서 이뤄지기 때문에 filter 조건을 판별하는 함수를 콜백으로 넘겨서 traverse 함수의 재사용성을 높이고자 했는데요! 탐색 조건에 부합하는지 판별하는 콜백함수 isMatchingNode 로 수정하겠습니다!

youzysu replied:
DOM 탐색을 여러번 다른 조건에서 이뤄지기 때문에 filter 조건을 판별하는 함수를 콜백으로 넘겨서 traverse 함수의 재사용성을 높이고자 했는데요! 탐색 조건에 부합하는지 판별하는 콜백함수의 의미를 담을 수 있도록 isMatchingNode 로 수정하겠습니다!

lolWK commented on domAPI.js:
```diff
@@ -0,0 +1,74 @@
+//DFSPreOrder
+export const domAPI = {
+  htmlNode: document.firstElementChild,
```
firstElementChild로 설정한 이유가 무엇인가요?

youzysu replied:
DOM 탐색을 할 때 처음에 default params document인데, document는 element가 아니라서, 첫 element인 htmlNode를 기본 시작점으로 설정하기 위해 사용했습니다.

youzysu replied:
DOM 탐색을 할 때 처음에 default params document인데, document는 element가 아니라서, 첫 element인 htmlNode를 기본 시작점으로 설정하기 위해 사용했습니다.

youzysu committed:
feat: 빨간 우체통 탐색 정보를 화면에 렌더링한다.

- 직접 구현한 DOM 탐색 API를 활용하여 우체통을 가진 마을을 탐색한다.
- 우체통을 가진 마을의 이름을 화면에 렌더링한다.
- merge sort 정렬 알고리즘을 활용하여 우체통을 크기순으로 정렬한 결과를 렌더링한다.

altmit committed:
feat: 빨간 우체통이 있는 마을의 테두리는 2초 뒤에 빨간색으로 변경된다.
    - setTimeout을 활용해서 지연 처리
    - Promise 패턴을 적용

altmit committed:
refactor: 우체통 노드 탐색 및 렌더링 결과 보완

- 마을 내의 우체통 노드 탐색 시 구현한 querySelector로 수정
- 우체통 크기도 함께 렌더링
- 마을 이름 추가 렌더링

altmit committed:
refactor: 함수 분리 및 파일 구조 구성

youzysu committed:
refactor: 우체통 마을 정보 객체에서 우체통 크기 데이터 타입 수정

- string에서 number로 수정

youzysu committed:
refactor: 불필요한 코드 삭제
